[WIP]  Add wmftoolsenforcer admission controller

yuvipanda opened the pull request:
Enforces that:

```
  - All namespaces have a RunAsUser annotation
  - Pods in a namespace will have RunAsUser set in all the containers
    to the RunAsUser annotation
```

This wouldn't ever be merged, see https://github.com/kubernetes/kubernetes/pull/16250#issuecomment-155553224 for rationale. This is here for review.

yuvipanda committed:
Add wmftoolsenforcer admission controller

Enforces that:

  - All namespaces have a RunAsUser annotation
  - Namespaced users can't touch annotations on their own namespace
  - Pods in a namespace will have RunAsUser set in all the containers
    to the RunAsUser annotation

ncdc commented:
@pmorie @pweil- @smarterclayton @derekwaynecarr

thockin commented:
ok to test

smarterclayton commented:
This is similar to things we have discussed and implement in OpenShift, although we prefer to tie uid to service account or namespace.  I think this is a very common topology, and for folks who don't want to use a pod security policy (to allow admins to define rules) this is a good fall back.

@pweil-, do we have a clear path for end users where the PSP proposal and types is a natural evolution of one of a set of opinionated simpler admission strategies?

smarterclayton commented:
One concern i have is that this might be a bit too opinionated about what uid means for most users.  @derekwaynecarr if we had a better mechanism for configuring code admission controllers we could make this more flexible.

yuvipanda commented:
@smarterclayton can you point me to OpenShift docs on how it does this? I'm ok with tying this to the namespace too, since we restrict each user to their own namespace.

I agree this is too opinionated, but not sure how to make it more configurable / other ways people want to do this.

pweil- commented:
> @pweil-, do we have a clear path for end users where the PSP proposal and types is a natural evolution of one of a set of opinionated simpler admission strategies?

Coupling PSP based admission with other admission strategies is doable.  There are two main components

1.  Ensure that the admission strategies setting values are before the policy enforcer
2.  Ensure the identity has access to a policy that would allow the values set by step 1

For this particular scenario the policy could be set up such that a strategy allows the UID of an individual user.  It would be very similar to how we do it for namespace allocated UID ranges.   In this case the admission controller would look for a particular type of strategy that prompts it to configure the strategy dynamically like we do in OS but passes in the uid value rather than a namespace allocated range.

pweil- commented:
@yuvipanda - here is the overview for how SCCs are used in OpenShift.  SCC is the grandfather to what is now PodSecurityPolicy in my proposal.  https://docs.openshift.org/latest/architecture/additional_concepts/authorization.html#security-context-constraints

In OpenShift we have certain strategies which may be configured without limiting settings (selinux, uid, and soon to have fsgroup and supplemental group support).  In OpenShift, namespaces are allocated values that prompt the admission controller to configure these dynamic strategies with the pre-allocated values.

smarterclayton commented:
I'm also asking - how do we organize and document choices on the pod
security spectrum so that admins know how to pick? This is an example of a
very reasonable simple strategy that a PSP based approach might be overkill
for.  Do we collect a number of such strategies and check them into the
tree?  How far should we bend PSP to make it fit into specific shapes (from
not at all to all the way)?

I kind of agree that "use user's uid" doesn't seem unreasonable as a PSP
user ID choice - just wanted to make sure we consider both parts.

On Oct 29, 2015, at 12:50 PM, Paul Weil <notifications@github.com> wrote:

@pweil- <https://github.com/pweil->, do we have a clear path for end users
where the PSP proposal and types is a natural evolution of one of a set of
opinionated simpler admission strategies?

Coupling PSP based admission with other admission strategies is doable.
There are two main components

   1. Ensure that the admission strategies setting values are before the
   policy enforcer
   2. Ensure the identity has access to a policy that would allow the
   values set by step 1

For this particular scenario the policy could be set up such that a
strategy allows the UID of an individual user. It would be very similar to
how we do it for namespace allocated UID ranges. In this case the admission
controller would look for a particular type of strategy that prompts it to
configure the strategy dynamically like we do in OS but passes in the uid
value rather than a namespace allocated range.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16250#issuecomment-152245927>
.

yuvipanda commented:
So a couple of questions:

1. Does this have a realistic chance of getting merged, or should I just use this as a stopgap until PSP or something similar is flexible enough? I personally think this is useful in and of itself and should be merged but am biased :D
2. Comments on the code itself? This is my first Go patch of any size...

derekwaynecarr commented:
@yuvipanda - are you trying to tie this behavior to the end-user that created the pod?  are not most of your pods created by internal kubernetes controllers on behalf of other users?

yuvipanda commented:
@derekwaynecarr hmm, if a user creates an RC and the RC creates the pod, doesn't the admission controller fire with appropriate user info (that is, the info of the user who created the RC) when the RC is creating the pod? I had assumed so...

smarterclayton commented:
No, unfortunately.  We generally use service accounts for this tracking -
users can select service accounts, but the service account has control.  We
assign uids to service accounts via a policy in Openshift for that reason.
That said, I do think this sort of attribution is important in the future -
the use case is valid.

Today, there is no way to know which user created anything.  There's a
discussion going on in https://github.com/kubernetes/kubernetes/pull/13925

But the challenge is that won't prevent another user from modifying the rc
to do something else.  However, it would be a step in that direction.

On Nov 1, 2015, at 2:01 AM, Yuvi Panda <notifications@github.com> wrote:

@derekwaynecarr <https://github.com/derekwaynecarr> hmm, if a user creates
an RC and the RC creates the pod, doesn't the admission controller fire
with appropriate user info (that is, the info of the user who created the
RC) when the RC is creating the pod? I had assumed so...

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16250#issuecomment-152801442>
.

yuvipanda commented:
Ugh, so that throws a monkey wrench into my plans. 

Ultimately we want to make sure that users can't read other users' files on NFS, which requires we enforce uid (since they're all on the same mount). Not sure how exactly to solve that now.

yuvipanda commented:
Hmm, perhaps I can force this a bit more, and move the admission controller to also look at RC submissions and modify the podspec? :)

caesarxuchao commented:
sub/

yuvipanda commented:
https://github.com/kubernetes/kubernetes/blob/master/docs/design/service_accounts.md needs some more filling out, specifically:

> TBD: how policy limits which users can make pods with which service accounts.

My current reading of it allows any user to specify any service account, thus not really providing much use for my use case.

caesarxuchao commented:
cc @erictune since Eric drafted the service account design doc.

erictune commented:
I've read this issue and have been thinking about it.  I think we should do something to accommodate @yuvipanda use case, but I do worry about accepting something that is too divergent from where we want to grow to.  So, I've been trying to gather thoughts on what that is.

I agree with @smarterclayton that serviceAccount and namespace are the two best places to attach policy about what unix uids a pod can act as.  Of the two, I think serviceAccount is more general and better as a goal for how we implement it in kubernetes -- but doing it via namespace may be perfectly reasonable for more opinionated instances of kubernetes (e.g. yuvipanda's case or OpenShift's case).

I think we end up wanting to put something like securityContextController inside of, or referenced by each serviceAccount.

I've tried to find a way to avoid the need for controller-managers to do impersonation, since the places where I've seen that pattern internally have been a source of irritation.  But, the alternatives are either subject to confused-deputy (e.g. access review done inside each controller) or are very difficult to implement (e.g. delegation of narrowly scoped authority by users to controllers).  And, discussing it with bgrant0607, I think the sources of irritation are not going to affect any by the largest enterprises.  So, I've come around to agreeing with @smarterclayton that we have to do impersonation.

Once I have a clearer picture in my head of how this impersonation should work, I'll come back and suggest something tactical for @yuvipanda problem.  @yuvipanda I know I said I might have a suggestion today, but I won't.  But I will be working on it.

@smarterclayton @pweil- interested to hear more of your thoughts on impersonation.

erictune commented:
Outcome of discussion at Kubecon:

1. @yuvipanda to put an annotation on the namespace like "wikimedia-tool-user-uid=1234" when he creates it. 
1. then he modifies this PR so that it just looks at the namespace of a pod, then looks up the namespace, gets the annotation, parses it, and then sets the RunAsUid of all incoming pods.  Pods created by the cluster admin in the kube-system namespace would have some other annotation allowing them to do whatever they want.
1. he carries this patch for a few months and hopefully we get Security Policy #7893 merged into 1.3
1. @erictune to make sure that we fix https://github.com/kubernetes/kubernetes/pull/13667 or the equivalent ASAP.

yuvipanda commented:
@erictune it also turns out that even if I'm a user namespaced only to namespace X, I can actually modify the annotations on that namespace. I guess I'll need to deny this too in the admission controller.

yuvipanda commented:
Ok, so I'm still going to update this PR so I can get it reviewed but there's no expectation of this getting merged :)

smarterclayton commented:
Two other notes - Brian and I and a few others discussed defaulting and how
that's technically the role of LimitRanger in general (to apply a default
to a pod resource).  We discussed having LimitRange evolve into PodPolicy
which is the more generic complement to PodSecurityPolicy - to apply simple
defaults to pods that are created.  Things that could be defaulted / ranged
per namespace:

* node selector
* pod labels
* pod annotations
* environment variables onto pods
* specific secrets
* generic security defaults (vs the more sophisticated defaulting in PSP)

Not a truly generic solution *except* it allows simple proscriptive control
like node selector to be applied by a resource a regular user may not have
access to change.

Needs more discussion and I will spawn a follow up issue off this @liggitt
@derekwaynecarr @bgrant0607


On Nov 10, 2015, at 12:17 PM, Eric Tune <notifications@github.com> wrote:

Outcome of discussion at Kubecon:

   1. @yuvipanda <https://github.com/yuvipanda> to put an annotation on the
   namespace like "wikimedia-tool-user-uid=1234" when he creates it.
   2. then he modifies this PR so that it just looks at the namespace of a
   pod, then looks up the namespace, gets the annotation, parses it, and then
   sets the RunAsUid of all incoming pods. Pods created by the cluster admin
   in the kube-system namespace would have some other annotation allowing them
   to do whatever they want.
   3. he carries this patch for a few months and hopefully we get Security
   Policy #7893 <https://github.com/kubernetes/kubernetes/pull/7893> merged
   into 1.3
   4. @erictune <https://github.com/erictune> to make sure that we fix
   #13667 <https://github.com/kubernetes/kubernetes/pull/13667> or the
   equivalent ASAP.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16250#issuecomment-155553224>
.

smarterclayton commented:
https://github.com/kubernetes/kubernetes/issues/17097


On Nov 10, 2015, at 9:59 PM, Clayton Coleman <ccoleman@redhat.com> wrote:

Two other notes - Brian and I and a few others discussed defaulting and how
that's technically the role of LimitRanger in general (to apply a default
to a pod resource).  We discussed having LimitRange evolve into PodPolicy
which is the more generic complement to PodSecurityPolicy - to apply simple
defaults to pods that are created.  Things that could be defaulted / ranged
per namespace:

* node selector
* pod labels
* pod annotations
* environment variables onto pods
* specific secrets
* generic security defaults (vs the more sophisticated defaulting in PSP)

Not a truly generic solution *except* it allows simple proscriptive control
like node selector to be applied by a resource a regular user may not have
access to change.

Needs more discussion and I will spawn a follow up issue off this @liggitt
@derekwaynecarr @bgrant0607


On Nov 10, 2015, at 12:17 PM, Eric Tune <notifications@github.com> wrote:

Outcome of discussion at Kubecon:

   1. @yuvipanda <https://github.com/yuvipanda> to put an annotation on the
   namespace like "wikimedia-tool-user-uid=1234" when he creates it.
   2. then he modifies this PR so that it just looks at the namespace of a
   pod, then looks up the namespace, gets the annotation, parses it, and then
   sets the RunAsUid of all incoming pods. Pods created by the cluster admin
   in the kube-system namespace would have some other annotation allowing them
   to do whatever they want.
   3. he carries this patch for a few months and hopefully we get Security
   Policy #7893 <https://github.com/kubernetes/kubernetes/pull/7893> merged
   into 1.3
   4. @erictune <https://github.com/erictune> to make sure that we fix
   #13667 <https://github.com/kubernetes/kubernetes/pull/13667> or the
   equivalent ASAP.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16250#issuecomment-155553224>
.

yuvipanda commented:
/me tags @pmorie and @erictune :)

bgrant0607 commented:
Issue for automatic application of labels: #15390

pmorie commented on plugin/pkg/admission/wmftoolsenforcer/admission.go:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	// HACK: This should probably be a separate AdmissionController
+	hackNamespace, ok := a.GetObject().(*api.Namespace)
```
Yep, as we discussed at kubecon, there are three different admission controllers here:

1.  One that prevents namespaces from being created without the uid annotation
2.  One that prevents namespaces from being updated
3.  One that prevents pods from being created without the uid annotation

pmorie replied:
Also, as we discussed, if you needed to check for multiple types a type-switch is the idiomatic way to do it in go.

erictune replied:
For item 2 on pmorie's list, I was thinking you would use the ABAC policy to prevent users from modifying their namespace, not an admission controller.  Something like:
```
{"user":"bob", "resource": "pods", "namespace": "bob"}`
{"user":"bob", "resource": "services", "namespace": "bob"}`
{"user":"bob", "resource": "replicationControllers", "namespace": "bob"}`
{"user":"bob", "resource": "endpoints", "namespace": "bob"}`
{"user":"bob", "resource": "events", "namespace": "bob"}`
{"user":"bob", "resource": "secrets", "namespace": "bob"}`
{"user":"bob", "resource": "serviceAccounts", "namespace": "bob"}`
```
With a new block like that added for each new user you add.
Would that work for you, @yuvipanda ?

My reasoning is:

- use ABAC when possible rather than annotations because it is a more general mechanism and likely to be supported by project going forward better than your patch
- there are going to be other resources that you want to block users from too, which don't fit quite as neatly into this PR.
- better to whitelist allowed resources than to blacklist them.  That way, when a new feature in Kube comes out, you as the admin make an affirmative decision whether to allow users to use it.


Elaborating on that second point I just made: namespace is not the only object that is in a user's namespace which you don't want the user to have write access to.  You also would not want to user to modify `limitRange` or `resourceQuota` or `persistentVolumes`, if you start to use those concepts.

erictune replied:
Correction: namespace is not in a namespace, so even if you had said this in the ABAC file:
```
{"user":"bob",  "namespace": "bob"}
```
I don't think the user should be able to modify his namespace object, or if so that is a bug.

smarterclayton replied:
We generally adopted this pattern in Openshift - creating a virtual resource "project" that offers a limited edit on the namespace (certain white listed annotations) while denying namespace edit to all but cluster admins.

yuvipanda replied:
@erictune I already had abac of the form
   {"user":"bob",  "namespace": "bob"}

and that let bob annotate namespace without denying. I'll verify this again and file a bug once I can confirm!

I've switched to the whitelist approach anyway, since you are right that does provide me more control and is what I want! I can remove the namespace modifier aspect of it now I think.

pmorie commented on plugin/pkg/admission/wmftoolsenforcer/admission_test.go:
```diff
@@ -0,0 +1,99 @@
+	}
+}
+
+func TestPodWithUID(t *testing.T) {
```
Another note from our convo at kubecon: table testing is the pattern frequently applied to keep overall numbers of test methods low.  You could refactor these tests to a single table test if you wanted.

erictune commented on plugin/pkg/admission/wmftoolsenforcer/admission.go:
```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
I know your present plan is not to merge this PR, but I am going to review it as if we were going to merge it, and if it happens that we converge on something we want to merge, so much the better.  And if you don't have time to reply to all comments, that is fine too.

erictune commented on plugin/pkg/admission/wmftoolsenforcer/admission.go:
```diff
@@ -0,0 +1,142 @@
+	store  cache.Store
+}
+
+// NewSecurityContextDeny creates a new instance of the SecurityContextDeny admission controller
```
comment above does not match function name,

erictune commented on plugin/pkg/admission/wmftoolsenforcer/admission.go:
```diff
@@ -0,0 +1,142 @@
+	}
+	namespace := namespaceObj.(*api.Namespace)
+
+	if _, ok := namespace.Annotations["RunAsUser"]; !ok {
```
Isn't this going to prevent you from managing pods in the kube-system namespace?  Or do you RunAsUser=something in that namespace too?

yuvipanda replied:
So right now I've nothing in the kube-system namespace (no working skydns yet), and you're right I'll have to figure something out for that. However, now that I've whitelisted the list of allowed resources even more maybe this entire thing can go.

yuvipanda replied:
Good point, I'm not particularly sure. Perhaps I can add another annotation ("System?") or just hardcode kube-system in there... Another option would be to actually give kube-system a uid in our LDAP system and have its RunAsUser set to that, but not sure how useful that'll be.

erictune commented on plugin/pkg/admission/wmftoolsenforcer/admission.go:
```diff
@@ -0,0 +1,142 @@
+					RunAsUser: &uid,
+				}
+			} else {
+				container.SecurityContext.RunAsUser = &uid
```
You should check if RunAsUser is nil.  If it is nil, then you should set it to uid.  If it is non-nil, you should reject the request outright, so the user knows that their explicit choice was invalid.

erictune commented on plugin/pkg/admission/wmftoolsenforcer/admission.go:
```diff
@@ -0,0 +1,142 @@
+		container := &pod.Spec.Containers[i]
+		uid, ok := strconv.ParseInt(namespace.Annotations["RunAsUser"], 10, 32)
+		if ok == nil {
+			if container.SecurityContext == nil {
```
I think there are some other things inside container.SecurityContext that are potentially dangerous, such as Capabilities.  You may want to reject any request that has a non-nil container.SecurityContext.

yuvipanda replied:
+1 great catch, thank you!

yuvipanda committed:
Stop checking for namespace modifications

Instead use ABAC to whitelist the resources that a namespaced
user can modify.

Also an ABAC line of form:
   { "user": "X", "namespace": "X" }

should not allow user X to modify namespace X itself, and if they
can it is a bug (that should be verified and reported)

yuvipanda committed:
Do not less non-empty SecurityContexts pass through

Since they can have other dangerous things (like Capabilities)

yuvipanda committed:
Fix comment to hide copypasta

yuvipanda commented:
I'll address more comments later tomorrow!

eparis commented:
@k8s-bot test this issue: #IGNORE

Tests have been pending for 24 hours
